tsl/src/compression/recompression.c

```diff
@@ -0,0 +1,1457 @@
+		 * should give us, how many following compressed batches to decompress.
+		 *
+		 * Based on the calculated value, we fetch next compressed batches and perform
+		 * recompression.
+		 */
+		total_uncompressed_rows_per_segment =
+			affectedsegments->num_matching_uncompressedrows[compressed_row_count];
+		unsigned int total_batches_to_decompress =
+			((total_uncompressed_rows_per_segment + metacount_val) /
+			 (10 * MAX_ROWS_PER_COMPRESSION));
+
+		CompressTuplesortstateCxt *tuplestorecxt = NULL;
+		tuplestorecxt = initialize_tuplestore(decompressor->out_rel, keys, n_keys);
+		uncompressed_slot = MakeTupleTableSlot(decompressor->out_desc, &TTSOpsMinimalTuple);
+		while (affectedsegments->num_matching_uncompressedrows[compressed_row_count] &&
+			   tuplesort_gettupleslot(uncompressed_tuplestorecxt->tuplestore,
+									  true /*=forward*/,
+									  false /*=copy*/,
+									  uncompressed_slot,
+									  NULL /*=abbrev*/))
+		{
+			tuplesort_puttupleslot(tuplestorecxt->tuplestore, uncompressed_slot);
+			affectedsegments->num_matching_uncompressedrows[compressed_row_count]--;
+		}
+		ExecDropSingleTupleTableSlot(uncompressed_slot);
+		if (next_tuple)
+		{
+			/* this compressed tuple might have already been decompressed */
+			ItemPointer ptr1 = &(next_tuple)->t_self;
+			ItemPointer ptr2 = &affectedsegments->compressed_tuples[compressed_row_count]->t_self;
+			if (ItemPointerCompare(ptr1, ptr2) != 0)
+			{
+				/* decompress affected batch */
+				row_compressor->sequence_num =
+					decompress_in_tuplestore(decompressor,
+											 affectedsegments
+												 ->compressed_tuples[compressed_row_count],
+											 tuplestorecxt->tuplestore);
+			}
+			else
+			{
+				heap_freetuple(next_tuple);
+				next_tuple = NULL;
+			}
+		}
+		else
+		{
+			/* decompress affected batch */
+			row_compressor->sequence_num =
+				decompress_in_tuplestore(decompressor,
+										 affectedsegments->compressed_tuples[compressed_row_count],
+										 tuplestorecxt->tuplestore);
+		}
+		while (total_batches_to_decompress)
+		{
+			/* fetch next batches which will be affected */
+			if (fetch_next_compressed_tuple(decompressor,
+											row_compressor,
+											uncompressed_chunk,
+											nsegmentby_cols,
+											compressed_slot,
+											&next_tuple))
+			{
+				/* decompress affected batch */
+				decompress_in_tuplestore(decompressor, next_tuple, tuplestorecxt->tuplestore);
+			}
+			total_batches_to_decompress--;
+		}
+		tuplesort_performsort(tuplestorecxt->tuplestore);
+		recompress_segment(tuplestorecxt->tuplestore, decompressor->out_rel, row_compressor);
+		cleanup_tuplestorecxt(tuplestorecxt);
+		compressed_row_count++;
+		affectedsegments->total_affected_compressed_rows--;
+	}
+	ExecDropSingleTupleTableSlot(compressed_slot);
+}
+
+static void
+do_legacy_recompression(RowDecompressor *decompressor, RowCompressor *row_compressor,
+						const CompressTuplesortstateCxt *uncompressed_tuplestorecxt)
+{
```

Comment:
Let's call it `do_full_recompression` or something like this, to better reflect what it does.